include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+ *
+ * Each sink registers with a (user-defined/exchangeable) global registry that can be
+ * queried by the non-GR caller to find the sink responsible for a given signal name, etc.
+ * and either retrieve a poller handler that allows asynchronous data from a different thread,
+ * or register a callback that is invoked by the sink if the user-conditions are met.
+ *
+ * <pre>
+ * @code
+ *         ╔═══════════════╗
+ *    in0 ━╢   data sink   ║                      ┌──── caller ────┐
+ * (err0) ━╢ (opt. error)  ║                      │                │
+ *         ║               ║  retrieve poller or  │ (custom non-GR │
+ *        ━╢ :signal_name  ║←--------------------→│  user code...) │
+ *        ━╢ :signal_unit  ║  register            │                │
+ *         ║ :...          ║  callback function   └───┬────────────┘
+ *         ╚═ GR block ═╤══╝                          │
+ *                      │                             │
+ *                      │                             │
+ *                      │      ╭─registry─╮           │
+ *            register/ │      ╞══════════╡           │ queries for specific
+ *          deregister  ╰─────→│ [sinks]  │←──────────╯ signal_info_t list/criteria
+ *                             ╞══════════╡
+ *                             ╰──────────╯
+ *
+ * </pre>
+ * Pollers can be configured to be blocking, i.e. blocks the flow-graph
+ * if data is not being retrieved in time, or non-blocking, i.e. data being dropped when
+ * the user-defined buffer size is full.
+ * N.B. due to the nature of the GR scheduler, signals from the same sink are notified
+ * synchronuously (/asynchronuously) if handled by the same (/different) sink block.
+ *
+ * @tparam T input sample type
+ */
+template<typename T>
+class data_sink : public node<data_sink<T>> {
+    struct abstract_listener;
+
+    static constexpr std::size_t                   _listener_buffer_size = 65536;
+    std::deque<std::unique_ptr<abstract_listener>> _listeners;
+    std::mutex                                     _listener_mutex;
+    gr::history_buffer<T>                          _history = gr::history_buffer<T>(1);
+
+public:
+    Annotated<float, "sample rate", Doc<"signal sample rate">, Unit<"Hz">>           sample_rate = 10000.f;
+    Annotated<std::string, "signal name", Visible>                                   signal_name;
+    Annotated<std::string, "signal unit", Visible, Doc<"signal's physical SI unit">> signal_unit;
+    Annotated<T, "signal min", Doc<"signal physical min. (e.g. DAQ) limit">>         signal_min;
+    Annotated<T, "signal max", Doc<"signal physical max. (e.g. DAQ) limit">>         signal_max;
+
+    IN<T, std::dynamic_extent, _listener_buffer_size>                                in;
+
+    struct poller {
+        gr::circular_buffer<T>            buffer       = gr::circular_buffer<T>(_listener_buffer_size);
+        decltype(buffer.new_reader())     reader       = buffer.new_reader();
```

Comment:
maybe re-use the existing infrastructure from `port<T>` we have there already definitions like:
 ```cpp
    using ReaderType    = decltype(std::declval<BufferType>().new_reader());
    using WriterType    = decltype(std::declval<BufferType>().new_writer());
    using IoType        = std::conditional_t<IS_INPUT, ReaderType, WriterType>;
    using TagReaderType = decltype(std::declval<TagBufferType>().new_reader());
    using TagWriterType = decltype(std::declval<TagBufferType>().new_writer());
    using TagIoType     = std::conditional_t<IS_INPUT, TagReaderType, TagWriterType>;

private:
    std::string  _name         = static_cast<std::string>(PortName);
    std::int16_t _priority     = 0; // → dependents of a higher-prio port should be scheduled first (Q: make this by order of ports?)
    std::size_t  _min_samples  = (MIN_SAMPLES == std::dynamic_extent ? 1 : MIN_SAMPLES);
    std::size_t  _max_samples  = MAX_SAMPLES;
    bool         _connected    = false;

    IoType       _ioHandler    = new_io_handler();
    TagIoType    _tagIoHandler = new_tag_io_handler();
 ```

Reply:
Hmm, that'd introduce 5 more `using` declarations that are then used exactly once, saving the `decltype()`. I find it more cryptic in the end, to find out what the type actually is.

Reply:
The idea is not to use the `using` but the whole `port<T>` declaration... there is a lot of stream/tag boiler plating that is duplicated here.

Reply:
Do you mean to add a port to poller, or...?